tiledb/sm/query/reader_base.cc

```diff
@@ -414,228 +414,200 @@ Status ReaderBase::read_tiles(
-  auto statuses = storage_manager_->io_tp()->wait_all_status(tasks);
-  for (const auto& st : statuses)
-    RETURN_CANCEL_OR_ERROR(st);
-
-  return Status::Ok();
-}
-
-Status ReaderBase::read_tiles(
-    const std::string& name,
-    const std::vector<ResultTile*>* result_tiles,
-    std::vector<ThreadPool::Task>* const tasks) const {
-  // Shortcut for empty tile vec
-  if (result_tiles->empty())
-    return Status::Ok();
-
-  // For each tile, read from its fragment.
-  const bool var_size = array_schema_->var_size(name);
-  const bool nullable = array_schema_->is_nullable(name);
+  // Populate the list of regions per file to be read.
+  std::map<URI, std::vector<std::tuple<uint64_t, void*, uint64_t>>> all_regions;
+  std::mutex all_regions_mutex;
 
-  // Gather the unique fragments indexes for which there are tiles
-  std::unordered_set<uint32_t> fragment_idxs_set;
-  for (const auto& tile : *result_tiles)
-    fragment_idxs_set.emplace(tile->frag_idx());
+  // Mutex protecting attr_tiles_ maps in all ResultTiles.
+  std::mutex attr_tiles_mtx;
 
-  // Put fragment indexes in a vector
-  std::vector<uint32_t> fragment_idxs_vec;
-  fragment_idxs_vec.reserve(fragment_idxs_set.size());
-  for (const auto& idx : fragment_idxs_set)
-    fragment_idxs_vec.emplace_back(idx);
+  // Run all tiles and attributes in parallel.
+  auto status = parallel_for_2d(
+      storage_manager_->compute_tp(),
+      0,
+      names->size(),
+      0,
+      result_tiles->size(),
+      [&](uint64_t name_idx, unsigned rt_idx) {
+        auto& name = names->at(name_idx);
+        auto tile = result_tiles->at(rt_idx);
 
-  // Protect all elements within `result_tiles`.
-  std::unique_lock<std::mutex> ul(result_tiles_mutex_);
+        // For each tile, read from its fragment.
+        const bool var_size = array_schema_->var_size(name);
+        const bool nullable = array_schema_->is_nullable(name);
 
-  // Populate the list of regions per file to be read.
-  std::map<URI, std::vector<std::tuple<uint64_t, void*, uint64_t>>> all_regions;
-  for (const auto& tile : *result_tiles) {
-    auto const fragment = fragment_metadata_[tile->frag_idx()];
-    const uint32_t format_version = fragment->format_version();
+        auto const fragment = fragment_metadata_[tile->frag_idx()];
+        const uint32_t format_version = fragment->format_version();
 
-    // Applicable for zipped coordinates only to versions < 5
-    if (name == constants::coords && format_version >= 5)
-      continue;
+        // Applicable for zipped coordinates only to versions < 5
+        if (name == constants::coords && format_version >= 5)
+          return Status::Ok();
 
-    // Applicable to separate coordinates only to versions >= 5
-    const bool is_dim = array_schema_->is_dim(name);
-    if (is_dim && format_version < 5)
-      continue;
+        // Applicable to separate coordinates only to versions >= 5
+        const bool is_dim = array_schema_->is_dim(name);
```

Comment:
This needs to be `fragment->array_schema()->is_dim` since the fragment's schema might be different than the latest one